### ISSUE TITLE
"no entry found for key" with simple example

### ISSUE THREAD

**Issue by inferrna** [zk4x/zyx#1]
```Rust
use zyx::{Tensor, DType, TensorSave};
use zyx_nn::{Linear, Module};

#[derive(Module)]
struct TinyNet {
    l0: Linear,
    l1: Linear,
    lr: f32,
}


fn main() {

    let mut net = TinyNet {
        l0: Linear::init(3, 1024, true, DType::F32).unwrap(),
        l1: Linear::init(1024, 2, true, DType::F32).unwrap(),
        lr: 0.0
    };

    let x = Tensor::from([2, 3, 1]).cast(DType::F32);
    let target = Tensor::from([5, 7]);

    // Zyx also provides some optimizers like SGD and Adam
    let mut optim = zyx_optim::SGD {
        learning_rate: 0.01,
        momentum: 0.9,
        nesterov: true,
        ..Default::default()
    };

    let train_steps = 100;
    for _ in 0..train_steps {
        let y = net.l0.forward(&x).unwrap().relu();
        let y = net.l1.forward(&y).unwrap().sigmoid();
        let loss = y.mse_loss(&target);

        let grads = loss.backward(&net);
        optim.update(&mut net, grads);

    }
}
```
Trace:

```
thread 'main' panicked at /home/inferno/.cargo/git/checkouts/zyx-ef5b220b8bfd8a6f/99fc140/zyx/src/graph.rs:161:19:
no entry found for key
stack backtrace:
   0: rust_begin_unwind
             at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/panicking.rs:74:14
   2: core::panicking::panic_display
             at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/panicking.rs:264:5
   3: core::option::expect_failed
             at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/option.rs:2021:5
   4: core::option::Option<T>::expect
             at /home/inferno/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:933:21
   5: <alloc::collections::btree::map::BTreeMap<K,V,A> as core::ops::index::Index<&Q>>::index
             at /home/inferno/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/collections/btree/map.rs:2363:23
   6: zyx::graph::Graph::axes
             at /home/inferno/.cargo/git/checkouts/zyx-ef5b220b8bfd8a6f/99fc140/zyx/src/graph.rs:161:19
   7: zyx::runtime::Runtime::backward
             at /home/inferno/.cargo/git/checkouts/zyx-ef5b220b8bfd8a6f/99fc140/zyx/src/runtime.rs:1084:32
   8: zyx::tensor::Tensor::backward
             at /home/inferno/.cargo/git/checkouts/zyx-ef5b220b8bfd8a6f/99fc140/zyx/src/tensor.rs:107:13
   9: my_zyx_test::main
             at ./src/main.rs:37:21
  10: core::ops::function::FnOnce::call_once
             at /home/inferno/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by Dzuchun** [comment]
I've come up with a minimal code that panics in a similar way

```rust
use zyx_cpu::DType;

mod nugraph2;

fn main() {
    let backend = zyx_cpu::device().unwrap();

    let x = backend.zeros([1], DType::F32);

    println!("{}", &x + 1);
    println!("{}", &x); // <-- panic occurs here
}
```

```
thread 'main' panicked at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-cpu-0.1.1/src/interpreter.rs:155:37:
no entry found for key
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/core/src/panicking.rs:76:14
   2: core::panicking::panic_display
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/core/src/panicking.rs:269:5
   3: core::option::expect_failed
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/core/src/option.rs:2018:5
   4: core::option::Option<T>::expect
             at [HOME]/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:933:21
   5: <alloc::collections::btree::map::BTreeMap<K,V,A> as core::ops::index::Index<&Q>>::index
             at [HOME]/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/collections/btree/map.rs:2373:23
   6: zyx_cpu::interpreter::Interpreter::get
             at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-cpu-0.1.1/src/interpreter.rs:155:37
   7: <zyx_cpu::interpreter::Interpreter as zyx_core::runtime::RuntimeBackend>::evaluate
             at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-cpu-0.1.1/src/interpreter.rs:43:30
   8: zyx_core::runtime::Runtime<R>::evaluate
             at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-core-0.1.1/src/runtime.rs:449:9
   9: zyx_core::runtime::Runtime<R>::load
             at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-core-0.1.1/src/runtime.rs:141:13
  10: <&zyx_cpu::CPU as zyx_core::backend::Backend>::load
             at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-cpu-0.1.1/src/lib.rs:161:9
  11: zyx_core::tensor::Tensor<B>::to_vec
             at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-core-0.1.1/src/tensor.rs:528:9
  12: <zyx_core::tensor::Tensor<B> as core::fmt::Display>::fmt
             at [HOME]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/zyx-core-0.1.1/src/tensor.rs:312:36
  13: core::fmt::rt::Argument::fmt
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/core/src/fmt/rt.rs:177:76
  14: core::fmt::write
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/core/src/fmt/mod.rs:1189:21
  15: std::io::Write::write_fmt
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/std/src/io/mod.rs:1884:15
  16: <&std::io::stdio::Stdout as std::io::Write>::write_fmt
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/std/src/io/stdio.rs:792:9
  17: <std::io::stdio::Stdout as std::io::Write>::write_fmt
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/std/src/io/stdio.rs:766:9
  18: std::io::stdio::print_to
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/std/src/io/stdio.rs:1122:21
  19: std::io::stdio::_print
             at /rustc/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/std/src/io/stdio.rs:1232:5
  20: cval::main
             at ./src/main.rs:11:5
  21: core::ops::function::FnOnce::call_once
             at [HOME]/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5

```

**Comment by zk4x** [comment]
I have fixed this, but the example still does not work, because gradients for some parameters are not getting calculated. The autograd engine is somewhat old and I have not tested it properly yet. Now I have added some tests and will work on it more. I have switched to using GradientTape in fashion similar to tensoflow, so I will have to rewrite the build_topo function in graph.rs.

The original version creates a set of requires_grad nodes when it is called. Now nodes that require_grad are listed in gradient_tape, so this will be used instead, making topo search faster.

The whole autograd is ~ 200 loc, the only complicated part is that topo search, so hopefully it won't take much time to write.

Thank you for your feedback.

**Comment by zk4x** [comment]
There was issue with handling of binary ops in kernel fusion.

Fixed by aa9d47cb2bf65eec1239b67f4a18b9d1ad859406
